### PR TITLE
[Work In Progress] execute file in terminal support

### PR DESCRIPTION
**Pull request by jabooth**
This PR is a work in progress that will implement #261.

It uses the yet-unreleased [terminal extension](https://github.com/Tyriar/vscode-terminal-api-example), and so can only be tested using an Insider build.

Right now this PR adds a new `Run Python file in Terminal` command to the palette that will run the focused editor window in the terminal using the `pythonPath` interpreter. 

It includes a grim `setTimeout` workaround due to there seemingly currently not being an API for finding out when a terminal is ready for use (see https://github.com/Microsoft/vscode/issues/10918).

It also doesn't include the right-click a file menu option - @DonJayamanne I couldn't find that extension point?

#### TODO

- [x] Remove `setTimeout`/`<any>` cast when final terminal API ships
- [x] Add right-click context menu support if possible
- [ ] reuse existing terminal menu (waiting on Microsoft/vscode#10925)
- [x] restrict command to only work on Python files

### CODE REVIEW

**Comment by DonJayamanne** [comment]
@jabooth , adding the explorer context menu is simple:
- package.json changes as follows:
   + [Sample (activation of extension)](https://github.com/DonJayamanne/gitHistoryVSCode/blob/master/package.json#L32)
   + [Sample (define command)](https://github.com/DonJayamanne/gitHistoryVSCode/blob/master/package.json#L55)
- Code changes (you will need to support Uri as the argument to your method )
   + [Sample - uri argument ](https://github.com/DonJayamanne/gitHistoryVSCode/blob/master/src/extension.ts#L20)

**Comment by DonJayamanne** [comment]
@jabooth , also added some inline comments

File: src/client/providers/execInTerminalProvider.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+                vscode.window.showErrorMessage('The active file needs to be saved before it can be run');
+                return;
+            } 
+        } else {
+            vscode.window.showErrorMessage('No open file to run in terminal');
+            return;
+        }
+    }
+
+    const terminal = (<any>vscode.window).createTerminal(`Ext Terminal`);
```
Comment (DonJayamanne) [inline src/client/providers/execInTerminalProvider.ts]: Please change title from 'Ext Terminal' to 'Python'

File: src/client/providers/execInTerminalProvider.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+function execInTerminal(filePath: string) {
+    const currentPythonPath = settings.PythonSettings.getInstance().pythonPath;
+
+    console.log(filePath);
+    if (filePath === undefined) {
+        const activeEditor = vscode.window.activeTextEditor;
+        if (activeEditor !== undefined) {
+            if (!activeEditor.document.isUntitled) {
+                filePath = activeEditor.document.fileName;
```
Comment (DonJayamanne) [inline src/client/providers/execInTerminalProvider.ts]: @jabooth  You might want to check whether the current file is a python file.
I believe you can inspect the language of the active editor.

**Comment by DonJayamanne** [comment]
@jabooth , brilliant work, thanks

**Comment by DonJayamanne** [comment]
@jabooth , would be great if we re-used the terminal window created, rather than creating a new one every time. 
Unfortunately there's a potential problem with that approach: https://github.com/Microsoft/vscode/issues/10925

**Comment by DonJayamanne** [comment]
@jabooth 
> reuse existing terminal menu (waiting on Microsoft/vscode#10925)  

This is the only item that pending and would like your feedback on how you think this should be resolved.

Consider the following scenario:
- You select a file and run it 
- Extension creates a new terminal and runs the code
- Once again you select a file and run it
- Do we create a new terminal or re-use the previously created one?

**Problem 1:**
- If the previously created terminal is still inside the python repl, for instance the code in the first python file stopped execution at an input statement.  
- Now trying to run the command 'python filename.py' will not work 
- Question: Do we instead create new terminals every time?  

**Problem 2:**  
- If we keep creating new terminals, we could end up with a large number of these?
- Who does the house keeping?
- Question: When running a file again, do we present the user with a prompt to close the previous terminal?

**Suggested workflow:**
- When a file is selected to be executed we check if a terminal was previously created
- If a previous terminal exists, then 
  + We prompt to close that and open in a new terminal?
  + We prompt to re-use this terminal?
- All other cases, create a new terminal?

What do you think?

**Comment by DonJayamanne** [comment]
Merged

**Comment by jabooth** [comment]
@DonJayamanne sorry I've had other commitments and not been able to get back to this yet. You sure it's OK to merge? Has the terminal API been merged into VScode?

**Pull request closed without merging**